polygeist/tools/cgeist/Lib/ValueCategory.cc

```diff
@@ -390,6 +400,45 @@ ValueCategory ValueCategory::IntCast(OpBuilder &Builder, Location Loc,
   return {Res, /*IsReference*/ false};
 }
 
+ValueCategory ValueCategory::PtrToInt(OpBuilder &Builder, Location Loc,
+                                      Type DestTy) const {
+  assert(val.getType().isa<LLVM::LLVMPointerType>() &&
+         "Expecting pointer source type");
+  assert(DestTy.isa<IntegerType>() &&
+         "Expecting floating point promotion type");
+
+  return Cast<LLVM::PtrToIntOp>(Builder, Loc, DestTy);
+}
+
+ValueCategory ValueCategory::IntToPtr(OpBuilder &Builder, Location Loc,
+                                      Type DestTy) const {
+  assert(val.getType().isa<IntegerType>() && "Expecting pointer source type");
+  assert(DestTy.isa<LLVM::LLVMPointerType>() &&
+         "Expecting floating point promotion type");
+
+  return Cast<LLVM::IntToPtrOp>(Builder, Loc, DestTy);
+}
+
+ValueCategory ValueCategory::BitCast(OpBuilder &Builder, Location Loc,
+                                     Type DestTy) const {
+  return Cast<LLVM::BitcastOp>(Builder, Loc, DestTy);
+}
+
+ValueCategory ValueCategory::MemRef2Ptr(OpBuilder &Builder,
+                                        Location Loc) const {
+  const auto Ty = val.getType().dyn_cast<MemRefType>();
+  if (!Ty) {
+    assert(val.getType().isa<LLVM::LLVMPointerType>() &&
```

Comment:
Why is it legal to call memref2pointer with a pointer?

Reply:
Following the philosophy of the rest of the casts we implemented, if the casting is a nop, we simply return `*this`. As there are lots of points in the code in which we can get either `memref` or pointers, this will save lots of lines checking whether we already have a pointer before performing this cast.